### ISSUE TITLE
Error while creating external links with unicode names

### ISSUE THREAD

**Issue by kormitigrov** [h5py/h5py#732]
We are unable to create an ExternalLink to the file with the unicode (namely cyrilic) name. A simple example follows:

```python
# -*- coding: utf-8 -*-

import h5py

# creating and opening file
h5file = h5py.File('d:/temp/test.h5', 'w')

# creating the link to the file with unicode name (existing)
# h5file['link'] = h5py.ExternalLink('d:/work/grp_results.h5', '/')
h5file['link'] = h5py.ExternalLink('d:/Работа/grp_results.h5', '/')

# accessing the link - does not work
print(h5file['link'])
```

The traceback is rather odd:

```
Traceback (most recent call last):
  File "C:\Program Files (x86)\JetBrains\PyCharm Community Edition 5.0.3\helpers\pydev\pydevd.py", line 2407, in <module>
    globals = debugger.run(setup['file'], None, None, is_module)
  File "C:\Program Files (x86)\JetBrains\PyCharm Community Edition 5.0.3\helpers\pydev\pydevd.py", line 1798, in run
    launch(file, globals, locals)  # execute the script
  File "C:\Program Files (x86)\JetBrains\PyCharm Community Edition 5.0.3\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc) 
  File "D:/REXLAB/test5.py", line 13, in <module>
    print(h5file['link'])
  File "h5py\_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-2.5.x\h5py\_objects.c:2467)
  File "h5py\_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-2.5.x\h5py\_objects.c:2424)
  File "C:\Program Files\WinPython-64bit-3.4.4.2Qt5\python-3.4.4.amd64\lib\site-packages\h5py\_hl\group.py", line 164, in __getitem__
    oid = h5o.open(self.id, self._e(name), lapl=self._lapl)
  File "h5py\_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-2.5.x\h5py\_objects.c:2467)
  File "h5py\_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-2.5.x\h5py\_objects.c:2424)
  File "h5py\h5o.pyx", line 190, in h5py.h5o.open (D:\Build\h5py\h5py-2.5.x\h5py\h5o.c:3516)
KeyError: "Unable to open object (Unable to open file: name = 'd:/temp\\grp_results.h5', errno = 2, error message = 'no such file or directory', flags = 1, o_flags = 2)"
```

The file d:/temp/test.h5 does not exist, the file d:/Работа/grp_results.h5 does exist. However, the traceback mentions d:/temp\\grp_results.h5 which is not mentioned anywhere in the code!

We're working with python 3.4.4, h5py 2.5.0 HDF5 1.8.15

**Comment by kormitigrov** [comment]
We have found a temporary workaround. In the file group.py in method Group.__setitem__ instead of 

>        elif isinstance(obj, ExternalLink):
>            self.id.links.create_external(name, self._e(obj.filename),
>                          self._e(obj.path), lcpl=lcpl, lapl=self._lapl)

we've put

>        elif isinstance(obj, ExternalLink):
>            import sys
>            self.id.links.create_external(name, obj.filename.encode(sys.getfilesystemencoding()),
>                          self._e(obj.path), lcpl=lcpl, lapl=self._lapl)

**Comment by dondublon** [comment]
Yep, we have the same problem. Pleeeeease, solve it!

**Comment by andreabedini** [comment]
@kormitigrov your workaround looks very close to be an actual solution. I'd go further and say that all `filename` parameters passed to HDF5 should be encoded with `sys.getfilesystemencoding`. Any opinion on this @andrewcollette ?

**Comment by andrewcollette** [comment]
I think using the local file system encoding is a good compromise.  The HDF5 library doesn't seem to specify an encoding, so I am confused to begin with about how external links are supposed to be cross-platform portable.  Right now the `self._e` encoding will generally use UTF-8 which evidently doesn't work with the Windows APIs that HDF5 is calling.

**Comment by andreabedini** [comment]
@andrewcollette HDF5 simply passes the filename to the OS. What encoding the OS expects is a blurry topic but we can assume we can use `sys.getfilesystemencoding()` given it's there for this exact reason [[docs](https://docs.python.org/3.5/library/sys.html#sys.getfilesystemencoding)].

@dondublon @kormitigrov What OS are you running on and what does `sys.getfilesystemencoding()` returns for you?

> how external links are supposed to be cross-platform portable

They are clearly not. I came to the conclusion HDF5 is oblivious to anything that has to do with string encodings. The bottom line seems to be 'everything works if you only use ASCII'.

I guess I can prepare a PR, but won't be too soon.

**Comment by aragilar** [comment]
It's better to use [os.fsencode](https://docs.python.org/3.6/library/os.html#os.fsencode) and [os.fsdecode](https://docs.python.org/3.6/library/os.html#os.fsdecode), as there are a number of corner-cases you have to work around. I backported them to python 2.6/2.7 in #716. I've added fixes for this, and I'm just writing the tests to be sure it works.

**Comment by aragilar** [comment]
#716 now contains fixes and tests for this.

**Comment by kormitigrov** [comment]
@andreabedini  we are working on win7_64bit (russian locale), sys.getfilesystemencoding() returns 'mbcs'.

**Comment by dondublon** [comment]
Yep, we use win7 x64.

**Comment by andreabedini** [comment]
@aragilar I wasn't aware of these recent developments :) I left some comments on the PR, thanks.
@kormitigrov @dondublon thanks!

**Comment by andreabedini** [comment]
Closing as this is fixed in #716, @kormitigrov @dondublon please test the main branch and feel free to reopen if the problem remains.
